aten/src/ATen/native/Linear.cpp

```diff
@@ -1,9 +1,202 @@
+
+  std::vector<int64_t> rpermutation(lro);
+  rpermutation.insert(rpermutation.end(), sum_dims_.begin(), sum_dims_.end());
+  rpermutation.insert(rpermutation.end(), ro.begin(), ro.end());
+  rpermutation.insert(rpermutation.end(), lo.begin(), lo.end());
+
+  std::vector<int64_t> opermutation(lro.size()+lo.size()+sum_dims_.size()+ro.size(), -1);
+  {
+  int64_t i = 0;
+
+  for (auto it = lro.begin(); it != lro.end(); i++, it++) {
+    opermutation[*it] = i;
+  }
+  for (auto it = lo.begin(); it != lo.end(); i++, it++) {
+    opermutation[*it] = i;
+  }
+  for (auto it = sum_dims_.begin(); it != sum_dims_.end(); i++, it++) {
+    opermutation[*it] = i;
+  }
+  for (auto it = ro.begin(); it != ro.end(); i++, it++) {
+    opermutation[*it] = i;
+  }
+  }
+
+  left = left.permute(lpermutation).reshape({lro_size, lo_size, sum_size});
+  right = right.permute(rpermutation).reshape({lro_size, sum_size, ro_size});
+  Tensor result = at::bmm(left, right);
+  result = result.view(out_size).permute(opermutation);
+  if (! keepdim) {
+    for (int i = dim-1; i>=0; i--)
+      if (sum_dims[i])
+	result.squeeze_(i);
+  }
+  return result;
+}
+
+
+Tensor einsum(String eqn, TensorList tensors) {
+  std::string in_eqn;
+  size_t pos;
+  std::vector<std::int64_t> number_of_occurences(26, 0);
+  std::vector<std::int64_t> last_occurence(26, -1);
+  std::vector<std::int64_t> sorted_position(26, -1);
+  if ((pos = eqn.find("->")) != std::string::npos) {
+    in_eqn = eqn.substr(0, pos);
+  } else {
+    in_eqn = eqn;
+  }
+
+  int64_t operand = 0;
+  std::stringstream eqn_stream(in_eqn);
+  std::string term;
+  while (! eqn_stream.eof()) {
+    std::getline(eqn_stream, term, ',');
+    int64_t dims_in_operand = 0;
+    for (auto &c : term) {
+      AT_ASSERT(('a' <= c) && (c <= 'z'), "only lowercase letters a-z allowed as indices");
+      int64_t index_num = c-'a';
+      number_of_occurences[index_num]++;
+      AT_ASSERT(last_occurence[index_num] < operand, "diagonals (multiple occurences of the same index for one tensor) not implemented yet")
+      last_occurence[index_num] = operand;
+      dims_in_operand++;
+    }
+    AT_ASSERT((int64_t) tensors.size()>operand, "more operands in equation than tensors");
+    AT_ASSERT(dims_in_operand == tensors[operand].dim(), "dimension mismatch for operand %zd: equation %zd, tensor %zd", operand, dims_in_operand, tensors[operand].dim());
+    operand++;
+  }
+  AT_ASSERT((int64_t) tensors.size()==operand, "more tensors than operands in equation");
+
+  int64_t num_outputs = 0;
+  std::vector<int64_t> position_labels;
+  if (pos != std::string::npos) {
+    for (auto &c : eqn.substr(pos+2)) {
+      AT_ASSERT(('a' <= c) && (c <= 'z'), "only lowercase letters a-z allowed as indices");
+      int64_t index_num = c-'a';
+      AT_ASSERT(sorted_position[index_num] == -1, "index %c occurs twice in output", c);
+      sorted_position[index_num] = num_outputs;
+      position_labels.push_back(index_num);
+      num_outputs++;
```

Comment:
There's only a single output, can you please rename this to `num_rhs_iterators`?

Reply:
`num_output_dims` or `output_dims` or so? It really is the number of dimensions in the output.